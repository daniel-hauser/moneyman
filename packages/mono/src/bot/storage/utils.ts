import { formatISO, parseISO, roundToNearestMinutes } from "date-fns";
import type { CompanyTypes } from "israeli-bank-scrapers";
import type { Transaction } from "israeli-bank-scrapers/lib/transactions.ts";

/**
 * Generates a hash for a transaction that can be used to ~uniquely identify it.
 * The hash is backwards compatible with the caspion hash.
 */
export function transactionHash(
  tx: Transaction,
  companyId: CompanyTypes,
  accountNumber: string,
) {
  const date = roundToNearestMinutes(parseISO(tx.date)).toISOString();
  const parts = [
    date,
    tx.chargedAmount,
    tx.description,
    tx.memo,
    companyId,
    accountNumber,
  ];

  return parts.map((p) => String(p ?? "")).join("_");
}

/**
 *
 * @param tx
 * @param companyId
 * @param accountNumber
 * @returns A unique id for a transaction
 */
export function transactionUniqueId(
  tx: Transaction,
  companyId: CompanyTypes,
  accountNumber: string,
) {
  const date = formatISO(tx.date, {
    representation: "date",
  });

  const parts = [
    date,
    companyId,
    accountNumber,
    tx.chargedAmount,
    tx.identifier || `${tx.description}_${tx.memo}`,
  ];
  return parts.map((p) => String(p ?? "").trim()).join("_");
}
