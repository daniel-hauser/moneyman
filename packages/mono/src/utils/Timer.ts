export class Timer {
  duration: number;
  startTime: number;

  constructor(public name: string) {
    this.startTime = performance.now();
  }

  end() {
    this.duration = performance.now() - this.startTime;
  }

  toString() {
    if (this.duration === undefined) {
      return this.name;
    }
    return `${this.name}, took ${(this.duration / 1000).toFixed(2)}s`;
  }
}
