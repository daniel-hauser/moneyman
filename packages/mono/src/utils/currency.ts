const symbolToCurrency = {
  $: "USD",
  "€": "EUR",
  "₪": "ILS",
  NIS: "ILS",
};

export function normalizeCurrency(currency: string | undefined) {
  if (!currency) return undefined;
  return (symbolToCurrency[currency] ?? currency).toUpperCase();
}
