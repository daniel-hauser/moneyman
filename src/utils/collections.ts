export function addToKeyedSet<K, V>(map: Map<K, Set<V>>, key: K, value: V) {
  if (!map.has(key)) {
    map.set(key, new Set());
  }
  map.get(key)!.add(value);
}

export function addToKeyedMap<K, KItem, VItem>(
  map: Map<K, Map<KItem, VItem>>,
  key: K,
  kv: [KItem, VItem],
) {
  if (!map.has(key)) {
    map.set(key, new Map());
  }
  map.get(key)!.set(...kv);
}
