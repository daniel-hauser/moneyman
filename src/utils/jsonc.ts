import * as jsonc from "jsonc-parser";

/**
 * Parse JSONC (JSON with Comments) string into a JavaScript object
 * @param jsoncString - The JSONC string to parse
 * @returns Parsed JavaScript object
 * @throws Error if parsing fails
 */
export function parseJsonc(jsoncString: string): unknown {
  const errors: jsonc.ParseError[] = [];
  const result = jsonc.parse(jsoncString, errors, {
    allowTrailingComma: true,
    disallowComments: false,
    allowEmptyContent: false,
  });
  if (errors.length > 0) {
    const errorMessages = errors.map((error) => {
      const { error: errorCode, offset, length } = error;
      const errorType = jsonc.printParseErrorCode(errorCode);
      const location = `at position ${offset}${length ? `-${offset + length}` : ""}`;
      return `${errorType} ${location}`;
    });
    throw new Error(`JSONC parsing failed: ${errorMessages.join(", ")}`);
  }

  return result;
}

/**
 * Parse a JSONC string that should represent a config object
 * @param jsoncString - The JSONC string to parse
 * @returns Parsed config object
 * @throws Error if parsing fails or result is not an object
 */
export function parseJsoncConfig(jsoncString: string): Record<string, unknown> {
  const result = parseJsonc(jsoncString);

  if (typeof result !== "object" || result === null || Array.isArray(result)) {
    throw new Error("JSONC config must be an object");
  }

  return result as Record<string, unknown>;
}
