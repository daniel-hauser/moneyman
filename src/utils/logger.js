// utils/logger.js
import debug from "debug";

// Create a logger instance
export const logger = debug("moneyman");

// Export the createLogger function
export function createLogger(name) {
  return logger.extend(name);
}

// Export the logToPublicLog function
export function logToPublicLog(message) {
  console.log(message);
}
