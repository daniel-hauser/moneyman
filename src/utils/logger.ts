import debug from "debug";

export const logger = debug("moneyman");

export function createLogger(name: string) {
  return logger.extend(name);
}

/**
 * Logs a message to stdout. Will be publicly visible when running in GitHub actions
 * @unsafe
 */
export function logToPublicLog(message: string) {
  console.log(message);
}

export const metadataLogEntries: string[] = [];
const metadataLogger = createLogger("metadataLogger");

/**
 * Logs a message to the metadata file sent to the chat
 * @param message The message to log.
 */
export function logToMetadataFile(message: string) {
  const date = new Date().toISOString();
  metadataLogEntries.push(`[${date}] ${message}`);
  metadataLogger(message);
}
