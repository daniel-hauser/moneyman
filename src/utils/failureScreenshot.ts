import os from "node:os";
import fs from "node:fs";
import path from "node:path";
import { glob } from "glob";
import { createLogger } from "./logger.js";
import type { ImageWithCaption } from "../types.js";

const logger = createLogger("failureScreenshot");
const tempFolder = path.join(fs.realpathSync(os.tmpdir()), "moneyman");

export function getFailureScreenShotPath(companyId: string) {
  const companyDir = path.join(tempFolder, companyId);
  if (!fs.existsSync(companyDir)) {
    fs.mkdirSync(companyDir, { recursive: true });
  }

  const filePath = path.join(companyDir, `${companyId}-${Date.now()}.png`);
  logger("getFailureScreenShotPath %o", filePath);

  return filePath;
}

export async function sendFailureScreenShots(
  sendPhotos: (photos: Array<ImageWithCaption>) => Promise<unknown>,
) {
  try {
    const photos = await glob(`${tempFolder}/**/*.png`, { absolute: true });
    logger("Sending failure screenshots", { photos });
    await sendPhotos(
      photos.map((photoPath) => ({
        photoPath,
        caption: `[${path.basename(path.dirname(photoPath))}] Failure screenshot`,
      })),
    );
    fs.rmSync(tempFolder, { recursive: true, force: true });
  } catch (e) {
    logger("Failed to send failure screenshots", e);
  }
}
