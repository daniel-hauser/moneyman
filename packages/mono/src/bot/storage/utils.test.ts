/* eslint-disable @typescript-eslint/no-explicit-any */
import type { Transaction } from "israeli-bank-scrapers/lib/transactions.js";
import {
  TransactionStatuses,
  TransactionTypes,
} from "israeli-bank-scrapers/lib/transactions.js";
import { transactionUniqueId, transactionHash } from "./utils.ts";
import { CompanyTypes } from "israeli-bank-scrapers";

const transaction1: Transaction = {
  date: "2021-01-01",
  processedDate: "2021-01-01",
  chargedAmount: 1000,
  description: "desc 1",
  memo: "memo 1",
  originalAmount: 1000,
  originalCurrency: "ILS",
  identifier: "identifier1",
  status: TransactionStatuses.Completed,
  type: TransactionTypes.Normal,
};

const transaction2: Transaction = {
  date: "2021-01-01",
  processedDate: "2021-01-01",
  chargedAmount: 999,
  description: "desc 2",
  memo: "memo 2",
  originalAmount: 999,
  originalCurrency: "ILS",
  identifier: "identifier2",
  status: TransactionStatuses.Completed,
  type: TransactionTypes.Normal,
};

describe("transactionHash", () => {
  it("should return the same hash for the same transaction", () => {
    const companyId = CompanyTypes.hapoalim;
    const accountNumber = "123";
    const hash1 = transactionHash(transaction1, companyId, accountNumber);
    const hash2 = transactionHash(transaction1, companyId, accountNumber);

    expect(hash1).toBe(hash2);
  });

  it("should return different hashes for different transactions", () => {
    const companyId = CompanyTypes.hapoalim;
    const accountNumber = "123";
    const hash1 = transactionHash(transaction1, companyId, accountNumber);
    const hash2 = transactionHash(transaction2, companyId, accountNumber);

    expect(hash1).not.toBe(hash2);
  });

  it("should return different hashes for different company ids", () => {
    const accountNumber = "123";
    const hash1 = transactionHash(
      transaction1,
      CompanyTypes.leumi,
      accountNumber,
    );

    const hash2 = transactionHash(
      transaction1,
      CompanyTypes.isracard,
      accountNumber,
    );

    expect(hash1).not.toBe(hash2);
  });

  it("should return different hashes for different account numbers", () => {
    const companyId = CompanyTypes.hapoalim;
    const hash1 = transactionHash(transaction1, companyId, "123");
    const hash2 = transactionHash(transaction1, companyId, "456");

    expect(hash1).not.toBe(hash2);
  });

  it("Issue #132: hash has no nullish string", () => {
    const transactionWithUndefined = {
      ...transaction1,
      chargedAmount: undefined,
      description: undefined,
      memo: undefined,
    };

    const transactionWithNull = {
      ...transaction1,
      chargedAmount: null,
      description: null,
      memo: null,
    };

    for (const transaction of [transactionWithNull, transactionWithUndefined]) {
      const hash = transactionHash(
        transaction as any,
        CompanyTypes.leumi,
        "123",
      );

      expect(hash).not.toContain("null");
      expect(hash).not.toContain("undefined");
    }
  });
});

describe("transactionUniqueId", () => {
  it("should return the same unique id for the same transaction", () => {
    const companyId = CompanyTypes.hapoalim;
    const accountNumber = "123";
    const id1 = transactionUniqueId(transaction1, companyId, accountNumber);
    const id2 = transactionUniqueId(transaction1, companyId, accountNumber);

    expect(id1).toBe(id2);
    expect(id1).toMatchSnapshot();
  });

  it("should return different unique ids for different transactions", () => {
    const companyId = CompanyTypes.hapoalim;
    const accountNumber = "123";
    const id1 = transactionUniqueId(transaction1, companyId, accountNumber);
    const id2 = transactionUniqueId(transaction2, companyId, accountNumber);

    expect(id1).not.toBe(id2);
  });

  it("should return different unique ids for different company ids", () => {
    const accountNumber = "123";
    const id1 = transactionUniqueId(
      transaction1,
      CompanyTypes.leumi,
      accountNumber,
    );

    const id2 = transactionUniqueId(
      transaction1,
      CompanyTypes.isracard,
      accountNumber,
    );

    expect(id1).not.toBe(id2);
  });

  it("should return different unique ids for different account numbers", () => {
    const companyId = CompanyTypes.hapoalim;
    const id1 = transactionUniqueId(transaction1, companyId, "123");
    const id2 = transactionUniqueId(transaction1, companyId, "456");

    expect(id1).not.toBe(id2);
  });

  it("should convert nullish values to empty strings", () => {
    const transactionWithUndefined = {
      ...transaction1,
      chargedAmount: undefined,
      description: undefined,
      memo: undefined,
    };

    const transactionWithNull = {
      ...transaction1,
      chargedAmount: null,
      description: null,
      memo: null,
    };

    for (const transaction of [transactionWithNull, transactionWithUndefined]) {
      const hash = transactionHash(
        transaction as any,
        CompanyTypes.leumi,
        "123",
      );

      expect(hash).not.toContain("null");
      expect(hash).not.toContain("undefined");
    }
  });

  it('should return different unique ids for transactions with different "identifier" values', () => {
    const companyId = CompanyTypes.hapoalim;
    const accountNumber = "123";
    const tx1 = { ...transaction1, identifier: "1" };
    const tx2 = { ...transaction1, identifier: "2" };

    // The old hash will be the same for both transactions because it doesn't include the identifier
    expect(transactionHash(tx1, companyId, accountNumber)).toBe(
      transactionHash(tx2, companyId, accountNumber),
    );

    // The unique id will be different because it includes the identifier
    expect(transactionUniqueId(tx1, companyId, accountNumber)).not.toBe(
      transactionUniqueId(tx2, companyId, accountNumber),
    );
  });

  it.each([undefined, null, ""])(
    'should fallback to "description" + "memo" if "identifier" is %o',
    (identifier) => {
      const tx = {
        ...transaction1,
        identifier,
      };
      const id = transactionUniqueId(tx as any, CompanyTypes.hapoalim, "123");
      expect(id).toContain(`${tx.description}_${tx.memo}`);
    },
  );
});
