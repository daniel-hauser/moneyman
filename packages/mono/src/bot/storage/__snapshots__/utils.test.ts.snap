// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transactionUniqueId should return the same unique id for the same transaction 1`] = `"2021-01-01_hapoalim_123_1000_identifier1"`;
